.github/workflows/ci.yaml

```diff
@@ -8,177 +8,303 @@ on:
+          key: yarn-build-${{ hashFiles('**/yarn.lock') }}
+
+      - name: Install dependencies
+        if: steps.cache-yarn.outputs.cache-hit != 'true'
+        run: yarn --frozen-lockfile
+
+      - name: Build code-server
+        run: yarn build
+
+      # Parse the hash of the latest commit inside lib/vscode
+      # use this to avoid rebuilding it if nothing changed
+      - name: Get latest lib/vscode rev
+        id: vscode-rev
+        run: echo "::set-output name=rev::$(git log -1 --format='%H' ./lib/vscode)"
+
+      - name: Attempt to fetch vscode build from cache
+        id: cache-vscode
+        uses: actions/cache@v2
+        with:
+          path: |
+            lib/vscode/.build
+            lib/vscode/out-build
+            lib/vscode/out-vscode
+            lib/vscode/out-vscode-min
+          key: vscode-build-${{ steps.vscode-rev.outputs.rev }}
+
+      - name: Build vscode
+        if: steps.cache-vscode.outputs.cache-hit != 'true'
+        run: yarn build:vscode
+
+      # The release package does not contain any native modules
+      # and is neutral to architecture/os/libc version.
+      - name: Create release package
+        run: yarn release
+        if: success()
+
+      # https://github.com/actions/upload-artifact/issues/38
+      - name: Compress release package
+        run: tar -czf package.tar.gz release
+
       - name: Upload npm package artifact
         uses: actions/upload-artifact@v2
         with:
           name: npm-package
-          path: ./release-npm-package
+          path: ./package.tar.gz
+
+  # We use Ubuntu 16.04 here, so that our build is more compatible
+  # with older libc versions. We used to (Q1'20) use CentOS 7 here,
+  # but it has a full update EOL of Q4'20 and a 'critical security'
+  # update EOL of 2024. We're dropping our full support 3 quarters
+  # early, but this gets us CI speed benefits on GH.
+
+  # TODO: cache building yarn --production
+  # possibly 2m30s of savings(?)
+  # this requires refactoring our release scripts
+
+  # TODO: cross-compile on arm64
+  package:
+    name: x86-64 builds
+    needs: build
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-16.04, macos-latest]
+      fail-fast: false
 
-  linux-amd64:
-    needs: release
-    runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
-      - name: Download npm package
-        uses: actions/download-artifact@v2
-        with:
-          name: npm-package
-          path: ./release-npm-package
-      - name: Run ./ci/steps/release-packages.sh
-        uses: ./ci/images/centos7
-        with:
-          args: ./ci/steps/release-packages.sh
-      - name: Upload release artifacts
-        uses: actions/upload-artifact@v2
+
+      - name: Install node.js v12
+        uses: actions/setup-node@v2
         with:
-          name: release-packages
-          path: ./release-packages
+          node-version: "12"
+
+      - name: Install nfpm
```

Comment:
What's this used for? I had a note to myself to see if we could remove it

Reply:
`nfpm` is what we're using right now to build `.deb`/`.rpm` files - it may be possible to move off nfpm, but I think that's better suited for another PR down the line :p

Reply:
Sounds good to me!